### DISCUSSION TITLE
Arity of language

### DISCUSSION THREAD

**Discussion by 0x0f0f0f** [egraphs-good/egg#65] [E-graph theory]
I'm reading the paper on the [Demoura matcher](http://leodemoura.github.io/files/ematching.pdf). Does it strongly require that the function symbols in a language have a fixed arity? Me and Phil Zucker based our pattern matcher for Metatheory.jl on top of the simpler HP Simplify matcher, which apparently has worked well with (implicitly defined) languages where the arity of function symbols can be variable. We are planning to move to this more performant e-matcher.

**Comment by mwillsey** [comment]
It does not. The current egg implementation supports variadic function symbols (see [docs](https://docs.rs/egg/0.6.0/egg/macro.define_language.html), last variant). I don't think it's really used anywhere, since you can't write variadic patterns.

We are working on a much faster e-matcher that will likely remove support for variadic function symbols.
